pkg/logger/logger_cluster.go

```diff
@@ -0,0 +1,156 @@
+}
+
+//GetChannelProvider returns the channel provider the Cluster wi
+func (s *simpleCluster) GetChannelProvider() ChannelProvider {
+	return s.channel
+}
+
+//StartListening enables the Cluster instance to listen to the channel found in Cluster#getChannelProvider
+//This will block all code executed, so use it in a go routine
+func (s *simpleCluster) StartListening() {
+	s.GetWaitGroup().Add(1)
+
+	for incomingMessage := range s.GetChannelProvider().GetChannel() {
+		if s.timedFlushNeeded() {
+			if s.isLoggerMessageCached(incomingMessage.GetLogger().GetID()) { //The message is not flushed this release
+				s.Flush()
+				s.resetFlushTimer()
+
+				s.appendMessage(incomingMessage)
+			} else { //The message is flushed in this release
+				s.appendMessage(incomingMessage)
+
+				s.Flush()
+				s.resetFlushTimer()
+			}
+		} else { //Time wise we don't need to flush
+			if s.isLoggerMessageCached(incomingMessage.GetLogger().GetID()) { //Overwriting logger message, gotta flush
+				s.Flush()
+				s.resetFlushTimer()
+			}
+
+			s.appendMessage(incomingMessage) //Add message to either this or new release, depending on prev if
+		}
+	}
+
+	s.Flush()
+	s.GetWaitGroup().Done()
+}
+
+//Flush flushes all cached messages to the pipeline
+func (s *simpleCluster) Flush() {
+	if len(s.messageCache) < 1 {
+		return
+	}
+
+	s.GetLoggerPipeline().Write(s.messageCache)
+	s.resetMessageCache()
+}
+
+//GetWaitGroup returns the waitgroup used to signal if the listening is running at the momement
+func (s *simpleCluster) GetWaitGroup() *sync.WaitGroup {
+	return s.waitGroup
+}
+
+//resetMessageCache simply clears the message cache
+func (s *simpleCluster) resetMessageCache() {
+	s.messageCache = nil
+	//signal the garbage collector to remove eventually cached array structure under it: https://stackoverflow.com/questions/16971741/how-do-you-clear-a-slice-in-go
+	//nil slices will still appedabe so we don't need to reallocate a slice instance using make
+}
+
+//isLoggerMessageCached returns if the logger id already has a cached ChannelMessage instance assigned
+func (s *simpleCluster) isLoggerMessageCached(loggerID int) bool {
+	for _, i := range s.messageCache {
+		if i.GetLogger().GetID() == loggerID {
+			return true
+		}
+	}
+	return false
+}
+
+//timedFlushNeeded returns if the cluster needs to flush its values due to the time limit being reached
+func (s *simpleCluster) timedFlushNeeded() bool {
+	return s.nextCacheFlush.Before(time.Now())
+}
+
+//resetFlushTimer resets the flush timer to now
+func (s *simpleCluster) resetFlushTimer() {
+	s.nextCacheFlush = time.Now().Add(s.clusterDuration)
+}
+
+//appendMessage adds a message to the message cache of the cluster
+func (s *simpleCluster) appendMessage(c ChannelMessage) {
+	s.messageCache = append(s.messageCache, c)
+}
+
+//NewLoggerCluster creates a new instance of the LoggerCluster inferface
+func NewLoggerCluster(pipeline Pipeline, channelProvider ChannelProvider, clusterDuration time.Duration) Cluster {
```

Comment:
same case as before, will be better to return the implementation(Struct), instead of the interface.